docs: misc

justinmk opened the pull request:
- ref #30415
- add help tags for common searches on https://neovim.io/
- fix #31327

justinmk committed:
docs: misc

fix #31327

dundargoc commented on runtime/doc/develop.txt:
```diff
@@ -300,10 +300,15 @@ vim.paste in runtime/lua/vim/_editor.lua like this: >
 
 See also |dev-naming|.
 
+- "A restriction at one semantic level translates to freedom and power at
+  another semantic level." https://blog.higher-order.com/blog/2014/12/21/maximally-powerful/
```
Also I don't even understand what the quote is supposed to mean. Who is the target audience here?

clason replied:
Certainly not mathematics, I can tell you that...

dundargoc replied:
Also I don't even understand what the quote is supposed to mean. Who is the target audience here?

justinmk replied:
> Are you using `develop.txt` as your personal bookmark or something? Wtf is this?

You quoted stuff I didn't put here. But yes, I put things in branches that may not make it into master.

This was some fumbling around trying to get some intuition of the "expression problem".

justinmk replied:
And this "personal bookmark" is something I put here in earnest to try to help others. So fuck off.

dundargoc replied:
> You quoted stuff I didn't put here.

I quoted stuff in the linked article, which I assumed one would need to read in order to understand the context of the quote.

> But yes, I put things in branches that may not make it into master.

How should I know which changes are "not for master" and which changes that are for master? I can only review the changes I see.

> This was some fumbling around trying to get some intuition of the "expression problem". Which seems to be the name for our constant struggles with API design.

Sure, but this seems to almost make things more confusing? I'm not sure what conclusions or guidance one should get from it. It kinda gives off the vibes that the design of neovim is left as an exercise to the reader.

> And this "personal bookmark" is something I put here in earnest to try to help others. So fuck off. 🖕

OK, didn't expect that reaction. I could've worded my confusion better, I apologize for how I phrased it.

I don't doubt you found it helpful. I found it unhelpful, pretentious, lacking in substance and ultimately that my time was wasted and am also earnestly trying to help people avoid the situation. I have said my peace, I will now fuck off 🫡

justinmk replied:
I removed it.

justinmk committed:
docs: help tags for common neovim.io searches

Problem:
Various keywords are commonly searched-for on https://neovim.io, but
don't have help tags.

Solution:
Add help tags.

justinmk commented on runtime/doc/intro.txt:
```diff
@@ -49,17 +49,19 @@ For more information try one of these:
+- Nvim home page: https://neovim.io/
+- Vim FAQ: https://vimhelp.org/vim_faq.txt.html
 
+			*download* *upgrade* *ubuntu*
+To install or upgrade Nvim, you can...
+- Download a pre-built archive:
+  https://github.com/neovim/neovim/releases
+- Use your system package manager:
+  https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-package
```
This is just a stub for now

MariaSolOs commented on runtime/doc/api.txt:
```diff
@@ -1125,6 +1125,19 @@ nvim_open_term({buffer}, {opts})                            *nvim_open_term()*
+        vim.api.nvim_create_user_command('TermHl', function()
+          local b = vim.api.nvim_create_buf(false, true)
+          local chan = vim.api.nvim_open_term(b, {})
+          vim.api.nvim_chan_send(chan, table.concat(
+            vim.api.nvim_buf_get_lines(0, 0, -1, false), '\n')
```
Nit: The misalignment of the `table.concat` parentheses makes it seem as if the parentheses were unbalanced. Having the entire `nvim_chan_send` in a single line "reads better" IMO.

MariaSolOs commented on runtime/doc/develop.txt:
```diff
@@ -434,7 +434,9 @@ Use existing common {verb} names (actions) if possible:
     - fmt:          Formats
-    - get:          Gets things (often by a query)
+    - get:          Gets things. Two variants (overloads):
+                    1. `get(id: int): any` returns one item.
+                    2. `get(filter: dict): any[]` returns a list.
```
Would it be an overkill to use a generic notation here to indicate that type specificity should be used (when possible)?

```suggestion
                    1. `get<T>(id: int): T` returns one item.
                    2. `get<T>(filter: dict): T[]` returns a list.
```

MariaSolOs commented on runtime/doc/gui.txt:
```diff
@@ -20,8 +20,9 @@ features, whereas help tags with the "ui-" prefix refer to the |ui-protocol|.
 (third-party) GUIs that you can use instead:
 
-- Firenvim (Nvim in your web browser!) https://github.com/glacambre/firenvim
+							*vscode*
```
Why add this header? It doesn't look like the list below is `vscode` exclusive.

justinmk replied:
It's a common search on https://neovim.io that currently yields no results.

For now this is just a "better than nothing" stub, but can revisit it later, or add more tags for the other items here, etc.

MariaSolOs commented on runtime/doc/intro.txt:
```diff
@@ -49,17 +49,19 @@ For more information try one of these:
+- Use your system package manager:
+  https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-package
 
 			*bugs* *bug-report*
```
Maybe we should also update this to:
```suggestion
			*bugs* *bug-report* *feature-requests*
```

MariaSolOs commented on runtime/doc/nvim.txt:
```diff
@@ -14,13 +14,14 @@ If you are new to Vim, try the 30-minute tutorial: >vim
-(especially editor and Vimscript features) is maintained where possible. See
-|vim-differences| for the complete reference of differences from Vim.
+(especially editor and Vimscript features, except |Vim9script|) is maintained
+where possible. See |vim-differences| for the complete reference of
```
"complete reference of differences" sounds a bit awkward IMO.

MariaSolOs commented on runtime/doc/nvim.txt:
```diff
@@ -14,13 +14,14 @@ If you are new to Vim, try the 30-minute tutorial: >vim
 
 ==============================================================================
-Transitioning from Vim                          *nvim-from-vim*
+Transitioning from Vim                                         *nvim-from-vim*
```
_just thinking out loud_ Wonder if one day it would make sense to add a "Transitioning from VSCode" section.

MariaSolOs commented on runtime/doc/nvim.txt:
```diff
@@ -70,5 +71,21 @@ the same Nvim configuration on all of your machines, by creating
+==============================================================================
+What next?                                                   *nvim-quickstart*
+
+If you are just trying out Vim/Nvim for a few minutes, and want to see the
```
The distros won't work for vanilla vim though.

justinmk committed:
fixup! docs: help tags for common neovim.io searches

brianhuster commented:
From `:h lua-eval`

> The (dual) equivalent of "vim.eval" for passing Lua values to Nvim is "luaeval".

Neovim doesn't have `vim.eval` for parsing Vimscript values to Lua, but only `nvim_eval` as Nvim API.

dundargoc commented:
> Should I open a PR for this?

Depends on "this", if you mean an actual `vim.eval` implementation then probably not. Maybe a help tag `vim.eval` in the section of `lua-eval` if we want to go with better discovery route.

brianhuster commented:
No, what I mean is just replacing `vim.eval` with `nvim_eval` in that sentence.
